packages/cli/lib/ModuleManager.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2020-2023 Delft University of Technology and SynTest contributors
+ *
+ * This file is part of SynTest Framework - SynTest Core Graphing Plugin.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as path from "path";
+import shell = require("shelljs");
+import globalModules = require("global-modules");
+import Yargs = require("yargs");
+import { PluginInterface } from "./moduleInterfaces/PluginInterface";
+import {
+  pluginAlreadyRegistered,
+  pluginNotFound,
+  pluginsNotFound,
+} from "./util/diagnostics";
+import { ApplicationInterface } from "./moduleInterfaces/ApplicationInterface";
+
+export class ModuleManager {
+  static instance: ModuleManager;
+
+  static initializeModuleManager() {
+    this.instance = new ModuleManager();
+  }
+
+  private _applications: Map<string, ApplicationInterface>;
+  // type -> name -> plugin
+  private _plugins: Map<string, Map<string, PluginInterface>>;
+
+  constructor() {
+    this._applications = new Map();
+    this._plugins = new Map();
+  }
+
+  get plugins() {
+    return this._plugins;
+  }
+
+  getPlugin(type: string, name: string) {
+    if (!this._plugins.has(type)) {
+      throw new Error(pluginNotFound(name, type));
+    }
+
+    if (!this._plugins.get(type).has(name)) {
+      throw new Error(pluginNotFound(name, type));
+    }
+
+    return this._plugins.get(type).get(name);
+  }
+
+  getPluginsOfType(type: string) {
+    if (!this._plugins.has(type)) {
+      throw new Error(pluginsNotFound(type));
+    }
+
+    return this._plugins.get(type);
+  }
+
+  async prepare() {
+    for (const application of this._applications.values()) {
+      if (application.prepare) {
+        await application.prepare();
+      }
+    }
+
+    for (const plugins of this.plugins.values()) {
+      for (const plugin of plugins.values()) {
+        if (plugin.prepare) {
+          await plugin.prepare();
+        }
+      }
+    }
+  }
+
+  async cleanup() {
+    for (const application of this._applications.values()) {
+      if (application.cleanup) {
+        await application.cleanup();
+      }
+    }
+
+    for (const plugins of this.plugins.values()) {
+      for (const plugin of plugins.values()) {
+        if (plugin.cleanup) {
+          await plugin.cleanup();
+        }
+      }
+    }
+  }
+
+  async configureApplicationCommands(yargs: Yargs.Argv) {
```

Comment:
We can drop one of the names here. Either `configureApplications` or `configureTools`. The fact that internally, this is called a command is probably not important.

Reply:
Fixed.